BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,194 @@
+**/
+VOID *
+EFIAPI
+FindS3Info (
+  IN  UINT8   Id
+  )
+{
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+  BOOLEAN           FoundInfo = FALSE;
+  BL_PLD_COMM_HDR   *CommHdr;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return NULL;
+  }
+
+  CommHdr = (BL_PLD_COMM_HDR *) SmmBase;
+  while ( ((UINT8 *)CommHdr < SmmBase + SmmSize) &&
+          CommHdr->Signature == BL_PLD_COMM_SIG) {
+    if (CommHdr->Id == Id) {
+      FoundInfo = TRUE;
+      break;
+    }
+    CommHdr = (BL_PLD_COMM_HDR *) ((UINT8 *)CommHdr + CommHdr->TotalSize);
+  }
+
+  if (!FoundInfo) {
+    return NULL;
+  }
+
+  return (VOID *) CommHdr;
+}
+
+
+typedef UINT32 (*REG_READ)  (UINTN);
+typedef VOID   (*REG_WRITE) (UINTN, UINT32);
+
+STATIC REG_WRITE  mRegWrite[2][3] = {
+  { (REG_WRITE) MmioWrite8, (REG_WRITE) MmioWrite16, (REG_WRITE) MmioWrite32 },
+  { (REG_WRITE) IoWrite8, (REG_WRITE) IoWrite16, (REG_WRITE) IoWrite32 }
+};
+
+STATIC REG_READ mRegRead[2][3] = {
+  { (REG_READ) MmioRead8, (REG_READ) MmioRead16, (REG_READ) MmioRead32 },
+  { (REG_READ) IoRead8, (REG_READ) IoRead16, (REG_READ) IoRead32 }
+};
+
+/**
+  Calculate the index in Reg_READ and REG_WRITE structs
+
+  @param    Width   Width enum as returned by payload
+
+  @retval   Count   Index required
+**/
+UINT8
+WidthIndex (
+  IN  UINT8   Width
+)
+{
+  UINT8   Count = 0;
+
+  while (Width) {
+    Width >>= 1;
+    Count++;
+  }
+
+  return Count - 1;
+}
+
+/**
+  This function restores the states of the registers that were
+  set to be saved by the bootloader in the normal boot path.
+  After the payload has updated the 'Val' field of the reg info
+  in the normal boot path, we simply OR the updated 'Val'
+  with the existing vale of the register in the S3 resume boot path.
+  This function is only called in the S3 resume path.
+
+  @param    S3SaveReg   S3_SAVE_REG info offset
+    
+  @retval   None
+
+**/
+VOID
+EFIAPI
+RestoreS3RegInfo (
+  IN  S3_SAVE_REG   *S3SaveReg
+  )
+{
+  UINT8                 Index;
+  UINT32                Data32;
+  UINT8                 Type;
+  UINT8                 Width;
+
+  if (S3SaveReg == NULL || S3SaveReg->S3SaveHdr.Id != S3_SAVE_REG_COMM_ID) {
+    DEBUG ((DEBUG_ERROR, "Invalid S3 Save Reg pointer\n"));
+    return;
+  }
+
+  for (Index = 0; Index < S3SaveReg->S3SaveHdr.Count; Index++) {
+    if (S3SaveReg->RegInfo[Index].Addr != 0x00) {
+      //
+      // Check register type and width
+      //
+      Type = S3SaveReg->RegInfo[Index].Type;
+      Width = S3SaveReg->RegInfo[Index].Width;
+      if ( ((Type != IO) && (Type != MEM)) ||
+           ((Width != WIDE32) && (Width != WIDE16)) ) {
+        DEBUG ((DEBUG_ERROR, "ERROR: not supported regtype (0x%x) or regwidth (0x%x)\n", Type, Width));
```

Comment:
since you do have error condition,  better to return Status  instead. 
In general we should not print error message too often since it comes with a size cost.  We can use status to return an error condition.  
Also since you are using array, better to compare with the array size to prevent access overflow.